### ISSUE TITLE
Problem: certain per-package client scripts are unnecessarily slow

### ISSUE THREAD

**Issue by sevein** [archivematica/Issues#315]
**Expected behaviour**
Client scripts that are executed once per package should perform faster!

**Current behaviour**
Testing with [46800-files.zip](https://github.com/artefactual/archivematica/files/2439480/46800-files.zip), I've noticed that some client scripts are extremely slow, e.g.:
- `Generate METS.xml document` took 2h11m
- `Sanitize object's file and directory names` took 3h13m
- `Add processed structMap to METS.xml document` ... many hours!

We can easily generate a report with the duration of these jobs that are executed only once per package.

**Steps to reproduce**
Transfer the dataset linked above. It has ~46,800 text files in it.

**Your environment (version of Archivematica, OS version, etc)**
AM18-dev

---
**For Artefactual use:**
Please make sure these steps are taken before moving this issue from Review to Verified in Waffle:

- All PRs related to this issue are properly linked 👍
- All PRs related to this issue have been merged 👍
- Test plan for this issue has been implemented and passed 👍
- Documentation regarding this issue has been written and it has been added to the release notes, if needed 👍

**Comment by sevein** [comment]
CC @cole.

**Comment by sevein** [comment]
With #1397 merged and all the work done in the new `create_transfer_mets` the behaviour now seen in qa/1.x is very different to the originally described. The new `Generate METS.xml document` (which also made `Add processed structMap to METS.xml document` redundant) took only 37 seconds to process a transfer with 10k text files, **compared to 3+ hours before the changes were applied**. 

`Sanitize object's file and directory names` is still very slow (and chatty), probably the only one left in transfer that could be easily improved - at least in the standard transfer type workflow. Perhaps also some parts in `Send to backlog`, in particular the search indexing (e.g. see `elasticSearchFunctions._index_transfer_files`).

**Comment by ross-spencer** [comment]
Nice work @cole and @sevein. For the sanitize names improvements I noticed Wellcome have some work there we might want to coordinate with, if that is in scope for `1.10` https://github.com/wellcometrust/archivematica/pull/2

**Comment by cole** [comment]
@sevein what system are you running that on!?! 💻 💨 I think I need a new laptop.

@ross-spencer good tip! Thanks. Looks like @alexwlchan has made some good changes there.

**Comment by sevein** [comment]
> @sevein what system are you running that on!?! computer dash I think I need a new laptop.

Linux 4.15 + i5-8400 +  DDR4-2666

**Comment by alexwlchan** [comment]
We’re planning to offer those patches upstream soon. :)

**Comment by cole** [comment]
@alexwlchan Thanks! I think in this case I'll cherry pick them into a branch, since I'm working on those files anyways.

**Comment by ross-spencer** [comment]
Sanitize names fixes may also (positively) impact https://github.com/archivematica/Issues/issues/352

**Comment by ablwr** [comment]
I tried testing this, but the above test data fails the Sanitize microservice - -testing on both 1.10RC CentOS and Xenial.

I could open a new issue, but I figure I should mention it here first.

**Comment by ablwr** [comment]
Here's what I've been getting on my local machine...


```
+----------+--------------------------------------------+-----------+-------+-----------------+
| Phase    | microserviceGroup                          | CPU Time  | Tasks | duration        |
+----------+--------------------------------------------+-----------+-------+-----------------+
| Transfer | Verify transfer compliance                 | 00:00:00  | 46866 | 00:00:58.706359 |
| Transfer | Rename with transfer UUID                  | 00:00:00  |     1 | 00:00:00.045215 |
| Transfer | Include default Transfer processingMCP.xml | 00:00:00  |     1 | 00:00:00.020199 |
| Transfer | Assign file UUIDs and checksums            | 26:05:34  | 93723 | 00:17:46.694477 |
| Transfer | Reformat metadata files                    | 00:00:00  |     1 | 00:00:00.020954 |
| Transfer | Verify transfer checksums                  | 00:00:00  |     1 | 00:00:00.029617 |
| Transfer | Quarantine                                 | 00:00:00  |     1 | 00:00:00.025980 |
| Transfer | Scan for viruses                           | 26:10:38  | 46863 | 00:18:01.854815 |
| Transfer | Generate transfer structure report         | 00:00:00  |     3 | 00:00:01.540876 |
| Transfer | Clean up names                             | 00:11:00  |     2 | 00:11:17.622581 |
| Transfer | Identify file format                       | 115:45:56 | 46862 | 13:58:45.827892 |
| Transfer | Extract packages                           | 00:04:06  |     2 | 00:04:09.090586 |
| Transfer | Characterize and extract metadata          | 159:58:16 | 46864 | 01:22:29.086781 |
| Transfer | Validation                                 | 12:27:41  | 46861 | 00:07:17.138585 |
| Transfer | Generate METS.xml document                 | 00:09:20  |     1 | 00:09:20.210149 |
| Transfer | Examine contents                           | 00:00:00  |     1 | 00:00:00.299769 |
| Transfer | Complete transfer                          | 00:00:00  |     3 | 00:00:01.104112 |
| Transfer | Create SIP from Transfer                   | 00:00:00  |     4 | 00:01:33.200496 |
+----------+--------------------------------------------+-----------+-------+-----------------+
```

**Comment by ablwr** [comment]
And for a basic test environment CentOS machine (the last being NULL because it's still churning through it or has gotten stuck/frozen -- it's been running overnight & more)

```
+----------+--------------------------------------------+-----------+-------+-----------------+
| Phase    | microserviceGroup                          | CPU Time  | Tasks | duration        |
+----------+--------------------------------------------+-----------+-------+-----------------+
| Transfer | Verify transfer compliance                 | 00:00:11  | 10006 | 00:00:20.878480 |
| Transfer | Rename with transfer UUID                  | 00:00:00  |     1 | 00:00:00.024317 |
| Transfer | Include default Transfer processingMCP.xml | 00:00:00  |     1 | 00:00:00.019075 |
| Transfer | Assign file UUIDs and checksums            | 05:37:33  | 20003 | 00:04:15.141726 |
| Transfer | Reformat metadata files                    | 00:00:00  |     1 | 00:00:00.010927 |
| Transfer | Verify transfer checksums                  | 00:00:00  |     1 | 00:00:00.023422 |
| Transfer | Quarantine                                 | 00:00:00  |     1 | 00:00:00.031149 |
| Transfer | Scan for viruses                           | 03:05:27  | 10003 | 00:02:23.684221 |
| Transfer | Generate transfer structure report         | 00:00:00  |     3 | 00:00:01.027558 |
| Transfer | Clean up names                             | 00:02:06  |     2 | 00:02:19.320846 |
| Transfer | Identify file format                       | 31:25:41  | 10002 | 00:21:05.991575 |
| Transfer | Extract packages                           | 00:00:24  |     2 | 00:00:25.417323 |
| Transfer | Characterize and extract metadata          | 838:59:59 | 10004 | 07:10:00.839480 |
| Transfer | Validation                                 | 00:04:39  | 10001 | 00:00:55.119799 |
| Transfer | Generate METS.xml document                 | NULL      |     1 | NULL            |
+----------+--------------------------------------------+-----------+-------+-----------------+
```

**Comment by sallain** [comment]
Moving this to done as testing has been successful, even though we don't have clear benchmarks to test against. More performance enhancements are forthcoming as well.
